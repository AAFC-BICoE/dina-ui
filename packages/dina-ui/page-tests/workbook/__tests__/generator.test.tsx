import { waitFor } from "@testing-library/react";
import { WorkbookTemplateGenerator } from "../../../pages/workbook/generator";
import { mountWithAppContext2 } from "../../../test-util/mock-app-context";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import {
  TEST_CLASSIFICATIONS,
  TEST_MANAGED_ATTRIBUTE_COLLECTING_EVENT,
  TEST_MANAGED_ATTRIBUTE_MATERIAL_SAMPLE,
  TEST_MANAGED_ATTRIBUTE_PREPARATION
} from "../__mocks__/generator.mock";

const mockPost = jest.fn();

const mockGet = jest.fn<any, any>(async (path, options) => {
  switch (path) {
    case "collection-api/managed-attribute":
      switch (options?.filter?.rsql) {
        case "managedAttributeComponent==MATERIAL_SAMPLE":
          return { data: [TEST_MANAGED_ATTRIBUTE_MATERIAL_SAMPLE] };
        case "managedAttributeComponent==PREPARATION":
          return { data: [TEST_MANAGED_ATTRIBUTE_PREPARATION] };
        case "managedAttributeComponent==COLLECTING_EVENT":
          return { data: [TEST_MANAGED_ATTRIBUTE_COLLECTING_EVENT] };
      }
    case "collection-api/vocabulary2/taxonomicRank":
      return { data: TEST_CLASSIFICATIONS };
  }
});

const apiContext: any = {
  apiClient: { get: mockGet, axios: { post: mockPost } }
};

describe("Workbook Template Generator", () => {
  beforeEach(() => jest.clearAllMocks());

  it("Page Layout", async () => {
    // Generates a snapshot of the generator page. This is used to ensure the design of the page
    // does not change unless intented.
    const wrapper = mountWithAppContext2(<WorkbookTemplateGenerator />, {
      apiContext
    });

    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  it("Select multiple columns and set aliases", async () => {
    const wrapper = mountWithAppContext2(<WorkbookTemplateGenerator />, {
      apiContext
    });

    // Click the "Add new column" dropdown
    userEvent.click(wrapper.getByRole("combobox"));
    await waitFor(() => {
      // Total number of options expected based on the dynamic config and index map returned.
      expect(wrapper.getAllByRole("option").length).toBeGreaterThanOrEqual(1);
    });

    // Select "Primary ID", give it alias of "Sample Name"
    userEvent.click(wrapper.getAllByRole("option", { name: /primary id/i })[0]);
    userEvent.click(wrapper.getAllByRole("button", { name: /add column/i })[0]);
    userEvent.type(
      wrapper.getAllByRole("textbox").at(-1) as HTMLElement,
      "Sample Name"
    );

    // Select "Barcode", give it alias of "Bar code"
    userEvent.click(wrapper.getByRole("combobox"));
    userEvent.click(wrapper.getByRole("option", { name: /barcode/i }));
    userEvent.click(wrapper.getByRole("button", { name: /add column/i }));
    userEvent.type(
      wrapper.getAllByRole("textbox").at(-1) as HTMLElement,
      "Bar code"
    );

    // Select a relationship level field, and give it an alias.
    userEvent.click(wrapper.getByRole("combobox"));
    userEvent.click(
      wrapper.getAllByRole("option", {
        name: /additional collection number/i
      })[0]
    );
    userEvent.click(wrapper.getByRole("button", { name: /add column/i }));
    userEvent.type(
      wrapper.getAllByRole("textbox").at(-1) as HTMLElement,
      "Coll number"
    );

    // Generate the template.
    userEvent.click(
      wrapper.getByRole("button", { name: /generate template/i })
    );
    await waitFor(() => {
      expect(mockPost).toHaveBeenCalledTimes(1);
    });
    expect(mockPost).toHaveBeenCalledWith(
      "objectstore-api/workbook/generation",
      {
        data: {
          attributes: {
            aliases: ["Sample Name", "Bar code", "Coll number"],
            columns: [
              "materialSampleName",
              "barcode",
              "collectingEvent.otherRecordNumbers"
            ]
          },
          type: "workbook-generation"
        }
      },
      {
        headers: {
          "Content-Type": "application/vnd.api+json"
        },
        responseType: "blob"
      }
    );
  });

  it("Delete and moving functionality in the generator", async () => {
    const wrapper = mountWithAppContext2(<WorkbookTemplateGenerator />, {
      apiContext
    });

    // Click the "Add new column" dropdown
    userEvent.click(wrapper.getByRole("combobox"));
    await waitFor(() => {
      // Total number of options expected based on the dynamic config and index map returned.
      expect(wrapper.getAllByRole("option").length).toBeGreaterThanOrEqual(1);
    });

    // Select "Primary ID", give it alias of "Sample Name"
    userEvent.click(wrapper.getAllByRole("option", { name: /primary id/i })[0]);
    userEvent.click(wrapper.getAllByRole("button", { name: /add column/i })[0]);
    await new Promise(setImmediate);
    userEvent.type(
      wrapper.getAllByRole("textbox").at(-1) as HTMLElement,
      "Sample Name"
    );

    // Select "Barcode", give it alias of "Bar code"
    userEvent.click(wrapper.getByRole("combobox"));
    userEvent.click(wrapper.getByRole("option", { name: /barcode/i }));
    userEvent.click(wrapper.getByRole("button", { name: /add column/i }));
    await new Promise(setImmediate);
    userEvent.type(
      wrapper.getAllByRole("textbox").at(-1) as HTMLElement,
      "Bar code"
    );

    // Select a relationship level field, and give it an alias.
    userEvent.click(wrapper.getByRole("combobox"));
    userEvent.click(
      wrapper.getAllByRole("option", {
        name: /additional collection number/i
      })[0]
    );
    userEvent.click(wrapper.getByRole("button", { name: /add column/i }));
    await new Promise(setImmediate);
    userEvent.type(
      wrapper.getAllByRole("textbox").at(-1) as HTMLElement,
      "Coll number"
    );

    // Remove the "Barcode" field.
    userEvent.click(wrapper.getAllByTestId("delete-button")[1]);

    // Move the material sample name down.
    userEvent.click(wrapper.getAllByTestId("move-up-button")[1]);

    // Generate the template.
    userEvent.click(
      wrapper.getByRole("button", { name: /generate template/i })
    );
    await waitFor(() => {
      expect(mockPost).toHaveBeenCalledTimes(1);
    });
    expect(mockPost).toHaveBeenCalledWith(
      "objectstore-api/workbook/generation",
      {
        data: {
          attributes: {
            aliases: ["Coll number", "Sample Name"],
            columns: [
              "collectingEvent.otherRecordNumbers",
              "materialSampleName"
            ]
          },
          type: "workbook-generation"
        }
      },
      {
        headers: {
          "Content-Type": "application/vnd.api+json"
        },
        responseType: "blob"
      }
    );
  });

  it("Select multiple columns and do not set aliases", async () => {
    const wrapper = mountWithAppContext2(<WorkbookTemplateGenerator />, {
      apiContext
    });

    // Click the "Add new column" dropdown
    userEvent.click(wrapper.getByRole("combobox"));
    await waitFor(() => {
      // Total number of options expected based on the dynamic config and index map returned.
      expect(wrapper.getAllByRole("option").length).toBeGreaterThanOrEqual(1);
    });

    // Select "Primary ID"
    userEvent.click(wrapper.getAllByRole("option", { name: /primary id/i })[0]);
    userEvent.click(wrapper.getAllByRole("button", { name: /add column/i })[0]);

    // Select "Barcode"
    userEvent.click(wrapper.getByRole("combobox"));
    userEvent.click(wrapper.getByRole("option", { name: /barcode/i }));
    userEvent.click(wrapper.getByRole("button", { name: /add column/i }));

    // Select a relationship level field
    userEvent.click(wrapper.getByRole("combobox"));
    userEvent.click(
      wrapper.getAllByRole("option", { name: /collection number/i })[1]
    );
    userEvent.click(wrapper.getByRole("button", { name: /add column/i }));

    // Generate the template.
    userEvent.click(
      wrapper.getByRole("button", { name: /generate template/i })
    );
    await waitFor(() => {
      expect(mockPost).toHaveBeenCalledTimes(1);
    });
    expect(mockPost).toHaveBeenCalledWith(
      "objectstore-api/workbook/generation",
      {
        data: {
          attributes: {
            columns: [
              "materialSampleName",
              "barcode",
              "collectingEvent.otherRecordNumbers"
            ],
            aliases: [
              "Primary ID",
              "Barcode",
              "Collecting Event Additional Collection Numbers"
            ]
          },
          type: "workbook-generation"
        }
      },
      {
        headers: {
          "Content-Type": "application/vnd.api+json"
        },
        responseType: "blob"
      }
    );
  });

  it("Selected managed attribute fields and generate template", async () => {
    const wrapper = mountWithAppContext2(<WorkbookTemplateGenerator />, {
      apiContext
    });
    await new Promise(setImmediate);

    // Click the "Add new column" dropdown
    userEvent.click(wrapper.getByRole("combobox"));
    await waitFor(() => {
      // Total number of options expected based on the dynamic config and index map returned.
      expect(wrapper.getAllByRole("option").length).toBeGreaterThanOrEqual(1);
    });

    // Select "Material Sample Managed Attributes".
    userEvent.click(
      wrapper.getAllByRole("option", {
        name: /managed attributes/i
      })[0]
    );
    await new Promise(setImmediate);

    // Select a managed attribute to generate.
    userEvent.click(wrapper.getAllByRole("combobox")[1]);
    await new Promise(setImmediate);
    userEvent.click(
      wrapper.getByRole("option", { name: /my test managed attribute/i })
    );
    userEvent.click(wrapper.getByRole("button", { name: /add column/i }));
    userEvent.type(
      wrapper.getAllByRole("textbox").at(-1) as HTMLElement,
      "Managed Attribute Alias"
    );

    // Click the "Add new column" dropdown
    userEvent.click(wrapper.getByRole("combobox"));
    await waitFor(() => {
      // Total number of options expected based on the dynamic config and index map returned.
      expect(wrapper.getAllByRole("option").length).toBeGreaterThanOrEqual(1);
    });

    // Select "Preparation Managed Attributes".
    userEvent.click(
      wrapper.getAllByRole("option", {
        name: /preparation managed attributes/i
      })[0]
    );
    await new Promise(setImmediate);

    // Select a managed attribute to generate.
    userEvent.click(wrapper.getAllByRole("combobox")[1]);
    await new Promise(setImmediate);
    userEvent.click(
      wrapper.getByRole("option", {
        name: /test preparation managed attribute/i
      })
    );
    userEvent.click(wrapper.getByRole("button", { name: /add column/i }));
    userEvent.type(
      wrapper.getAllByRole("textbox").at(-1) as HTMLElement,
      "Another Managed Attribute"
    );

    // Click the "Add new column" dropdown
    userEvent.click(wrapper.getByRole("combobox"));
    await waitFor(() => {
      // Total number of options expected based on the dynamic config and index map returned.
      expect(wrapper.getAllByRole("option").length).toBeGreaterThanOrEqual(1);
    });

    // Select "Collecting Event" managed attribute
    userEvent.click(
      wrapper.getAllByRole("option", {
        name: /managed attributes/i
      })[2]
    );
    await new Promise(setImmediate);

    // Select a managed attribute to generate.
    userEvent.click(wrapper.getAllByRole("combobox")[1]);
    await new Promise(setImmediate);
    userEvent.click(
      wrapper.getByRole("option", {
        name: /test collecting event managed attribute/i
      })
    );
    userEvent.click(wrapper.getByRole("button", { name: /add column/i }));

    // Generate the template.
    userEvent.click(
      wrapper.getByRole("button", { name: /generate template/i })
    );
    await waitFor(() => {
      expect(mockPost).toHaveBeenCalledTimes(1);
    });
    expect(mockPost).toHaveBeenCalledWith(
      "objectstore-api/workbook/generation",
      {
        data: {
          attributes: {
            aliases: [
              "Managed Attribute Alias",
              "Another Managed Attribute",
              // No alias was defined, so the column label is used instead:
              "Test Collecting Event Managed Attribute"
            ],
            columns: [
              "managedAttributes.my_test_managed_attribute",
              "preparationManagedAttributes.test_preparation_managed_attribute",
              "collectingEvent.managedAttributes.test_collecting_event_managed_attribute"
            ]
          },
          type: "workbook-generation"
        }
      },
      {
        headers: {
          "Content-Type": "application/vnd.api+json"
        },
        responseType: "blob"
      }
    );
  });

  it("Selected scientificNameDetails fields and generate template", async () => {
    const wrapper = mountWithAppContext2(<WorkbookTemplateGenerator />, {
      apiContext
    });
    await new Promise(setImmediate);

    // Go through all the possible classifications from the mock.
    for (const element of TEST_CLASSIFICATIONS.vocabularyElements as any) {
      // Click the "Add new column" dropdown
      userEvent.click(wrapper.getByRole("combobox"));
      await waitFor(() => {
        // Total number of options expected based on the dynamic config and index map returned.
        expect(wrapper.getAllByRole("option").length).toBeGreaterThanOrEqual(1);
      });

      // Click the "Scientific Name Details" option.
      userEvent.click(
        wrapper.getByRole("option", {
          name: /scientific name details/i
        })
      );
      await new Promise(setImmediate);

      // A new dropdown should appear:
      expect(
        wrapper.getByText(/select classification rank\.\.\./i)
      ).toBeInTheDocument();
      userEvent.click(wrapper.getAllByRole("combobox")[1]);

      // Select classification name.
      userEvent.click(wrapper.getByRole("option", { name: element.name }));

      // Add the column.
      userEvent.click(wrapper.getByRole("button", { name: /add column/i }));
    }

    // Change one of the headers to make sure the alias is kept.
    userEvent.type(wrapper.getByPlaceholderText(/kingdom/i), "Kingdom Test");

    // Generate the template.
    userEvent.click(
      wrapper.getByRole("button", { name: /generate template/i })
    );
    await waitFor(() => {
      expect(mockPost).toHaveBeenCalledTimes(1);
    });

    // Ensure the request is correct.
    expect(mockPost).toHaveBeenCalledWith(
      "objectstore-api/workbook/generation",
      {
        data: {
          attributes: {
            aliases: [
              "Kingdom Test",
              "Phylum",
              "Class",
              "Order",
              "Family",
              "Genus",
              "Species",
              "Subspecies",
              "Variety"
            ],
            columns: [
              "Kingdom",
              "Phylum",
              "Class",
              "Order",
              "Family",
              "Genus",
              "Species",
              "Subspecies",
              "Variety"
            ]
          },
          type: "workbook-generation"
        }
      },
      {
        headers: {
          "Content-Type": "application/vnd.api+json"
        },
        responseType: "blob"
      }
    );
  });

  it("Template name validation", async () => {
    const wrapper = mountWithAppContext2(<WorkbookTemplateGenerator />, {
      apiContext
    });

    const templateNameInput = wrapper
      .getAllByRole("textbox")
      .at(0) as HTMLElement;

    // Put an invalid template name
    userEvent.type(templateNameInput, "Test.xlsx");

    // Click the "Add new column" dropdown
    userEvent.click(wrapper.getByRole("combobox"));
    await waitFor(() => {
      // Total number of options expected based on the dynamic config and index map returned.
      expect(wrapper.getAllByRole("option").length).toBeGreaterThanOrEqual(1);
    });

    // Select "Primary ID", give it alias of "Sample Name"
    userEvent.click(wrapper.getAllByRole("option", { name: /primary id/i })[0]);
    userEvent.click(wrapper.getAllByRole("button", { name: /add column/i })[0]);

    // After setting a column, the filename should still be there.
    expect(templateNameInput).toHaveDisplayValue("Test.xlsx");

    mockPost.mockReturnValue("pretendFileData");
    userEvent.click(
      wrapper.getByRole("button", { name: /generate template/i })
    );

    // Error should be displayed on the page.
    expect(
      wrapper.getByText(
        /please enter a valid filename\. only letters, numbers, spaces, hyphens, and underscores are allowed\./i
      )
    ).toBeInTheDocument();

    // Ensure the name is still displayed for the user to correct it.
    expect(templateNameInput).toHaveDisplayValue("Test.xlsx");
  });
});
