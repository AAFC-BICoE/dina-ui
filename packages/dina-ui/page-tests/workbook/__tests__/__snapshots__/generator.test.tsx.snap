// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workbook Template Generator Page Layout 1`] = `
<DocumentFragment>
  <div>
    <form
      action="#"
    >
      <div
        class="error-viewer"
      />
      <nav
        class="d-flex flex-column align-items-center"
        id="wb-tphp"
      >
        <a
          class="wb-link-inv wb-sl"
          href="#wb-cont"
        >
          Skip to Main Content
        </a>
        <a
          class="wb-link-inv wb-sl"
          href="#wb-info"
        >
          Skip to "About this site"
        </a>
        <a
          class="wb-link-inv wb-sl"
          href="?wbdisable=true"
          rel="alternate"
        >
          Switch to basic HTML version
        </a>
      </nav>
      <header>
        <div
          class="centered container-fluid"
        >
          <div
            class="header-container row d-flex  row row-cols-md-2 row-cols-1"
          >
            <div
              class="col"
            >
              <img
                alt=""
                class="logo"
                property="logo"
                src="/static/images/canadaLogo_en.svg"
              />
              <span
                class="wb-inv"
                property="name"
              >
                Government of Canada / 
                <span
                  lang="fr"
                >
                  Gouvernement du Canada
                </span>
              </span>
              <meta
                content="Canada"
                property="areaServed"
                typeof="Country"
              />
            </div>
            <div
              class="text-end  col"
            >
              <ul
                class="list-inline mt-1 mb-1"
              >
                <li
                  class="list-inline-item mr-1 my-auto"
                >
                  <a
                    class="px-0 btn btn-link"
                    href="/feedback"
                    role="button"
                    tabindex="0"
                  >
                    Feedback
                  </a>
                </li>
                <li
                  class="list-inline-item px-2"
                >
                  –
                </li>
                <li
                  class="list-inline-item mr-0"
                >
                  <div
                    data-testid="languageSelector"
                  >
                    <button
                      class="px-0 btn btn-link"
                      lang="fr"
                      type="button"
                    >
                      Français
                    </button>
                  </div>
                </li>
              </ul>
              <ul
                class="list-inline"
              >
                <li
                  class="list-inline-item my-auto"
                >
                  <div
                    class="d-flex align-items-center text-end"
                  >
                    <span
                      class="me-2 my-auto h4"
                    >
                      <svg
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 496 512"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 96c48.6 0 88 39.4 88 88s-39.4 88-88 88-88-39.4-88-88 39.4-88 88-88zm0 344c-58.7 0-111.3-26.6-146.5-68.2 18.8-35.4 55.6-59.8 98.5-59.8 2.4 0 4.8.4 7.1 1.1 13 4.2 26.6 6.9 40.9 6.9 14.3 0 28-2.7 40.9-6.9 2.3-.7 4.7-1.1 7.1-1.1 42.9 0 79.7 24.4 98.5 59.8C359.3 421.4 306.7 448 248 448z"
                        />
                      </svg>
                    </span>
                    <span
                      class="me-4 my-auto h5"
                    >
                      <a
                        href="/dina-user/view?id=&hideBackButton=true"
                      >
                        test-user
                      </a>
                    </span>
                    <button
                      class="btn btn-info logout-button my-auto"
                      type="button"
                    >
                      Logout
                    </button>
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </div>
        <nav
          class="app-bar navbar navbar-expand-lg navbar-light"
        >
          <div
            class="centered container-fluid"
          >
            <a
              class="app-name navbar-brand"
              href="/"
            >
              DINA (developer)
            </a>
            <button
              aria-controls="basic-navbar-nav"
              aria-label="Toggle navigation"
              class="navbar-toggler collapsed"
              type="button"
            >
              <span
                class="navbar-toggler-icon"
              />
            </button>
            <div
              aria-expanded="false"
              class="navbar-collapse collapse"
              id="basic-navbar-nav"
              role="menubar"
            >
              <div
                class="nav-item dropdown"
                role="menuitem"
              >
                <a
                  aria-expanded="false"
                  class="dropdown-toggle nav-link"
                  href="#"
                  role="button"
                  tabindex="0"
                >
                  Collection
                </a>
              </div>
              <div
                class="nav-item dropdown"
                role="menuitem"
              >
                <a
                  aria-expanded="false"
                  class="dropdown-toggle nav-link"
                  href="#"
                  role="button"
                  tabindex="0"
                >
                  Transactions
                </a>
              </div>
              <div
                class="nav-item dropdown"
                role="menuitem"
              >
                <a
                  aria-expanded="false"
                  class="dropdown-toggle nav-link"
                  href="#"
                  role="button"
                  tabindex="0"
                >
                  Object Store
                </a>
              </div>
              <div
                class="nav-item dropdown"
                role="menuitem"
              >
                <a
                  aria-expanded="false"
                  class="dropdown-toggle nav-link"
                  href="#"
                  role="button"
                  tabindex="0"
                >
                  Agents
                </a>
              </div>
              <div
                class="nav-item dropdown"
                role="menuitem"
              >
                <a
                  aria-expanded="false"
                  class="dropdown-toggle nav-link"
                  href="#"
                  role="button"
                  tabindex="0"
                >
                  Sequencing
                </a>
              </div>
              <div
                class="nav-item dropdown"
                role="menuitem"
              >
                <a
                  aria-expanded="false"
                  class="dropdown-toggle nav-link"
                  href="#"
                  role="button"
                  tabindex="0"
                >
                  Controlled Vocabulary
                </a>
              </div>
              <div
                class="nav-item dropdown"
                role="menuitem"
              >
                <a
                  aria-expanded="false"
                  class="dropdown-toggle nav-link"
                  href="#"
                  role="button"
                  tabindex="0"
                >
                  Configuration
                </a>
              </div>
            </div>
          </div>
        </nav>
      </header>
      <div
        class="button-bar button-bar-sticky undefined"
      >
        <div
          class="container-fluid centered"
        >
          <div
            class="d-flex row"
          >
            <div
              class="col-md-6 col-sm-12 mt-2"
            >
              <a
                class="back-button my-auto"
                href="/workbook/upload"
              >
                Back to Upload Workbook
              </a>
            </div>
            <div
              class="col-md-6 col-sm-12 d-flex"
            >
              <button
                class="btn ms-auto btn-primary"
                disabled=""
                style="width: 12rem;"
                type="submit"
              >
                Generate Template
              </button>
            </div>
          </div>
        </div>
      </div>
      <main
        class="container-fluid"
        role="main"
      >
        <h1
          id="wb-cont"
        >
          Generate Workbook Template
        </h1>
        <h4
          class="mt-4"
        >
          Settings
        </h4>
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <div
              class="list-inline d-flex flex-row gap-4 pt-2"
            >
              <div
                class="flex-grow-1"
              >
                <strong>
                  Template Name
                </strong>
                <div
                  class="mt-2 input-group"
                >
                  <input
                    aria-describedby="extension"
                    class="flex-grow-1 form-control form-control"
                    name="name"
                    placeholder="template"
                    value=""
                  />
                  <span
                    class="input-group-text"
                    id="extension"
                  >
                    .xlsx
                  </span>
                </div>
              </div>
              <div
                class="flex-grow-1"
              >
                <label
                  class="type-field mb-2 w-100 mb-3"
                >
                  <div
                    class="field-label label-col mb-2"
                  >
                    <div
                      class="d-flex align-items-center w-100"
                    >
                      <strong
                        class="me-2"
                      >
                        <div
                          class="type-field-header"
                        >
                          Type
                        </div>
                      </strong>
                    </div>
                  </div>
                  <div
                    class="field-col"
                    style="cursor: auto;"
                  >
                    <div
                      class=" css-3iigni-container"
                    >
                      <span
                        class="css-1f43avz-a11yText-A11yText"
                        id="react-select-2-live-region"
                      />
                      <span
                        aria-atomic="false"
                        aria-live="polite"
                        aria-relevant="additions text"
                        class="css-1f43avz-a11yText-A11yText"
                        role="log"
                      />
                      <div
                        aria-disabled="true"
                        class=" css-16xfy0z-control"
                      >
                        <div
                          class=" css-1fdsijx-ValueContainer"
                        >
                          <div
                            class=" css-olqui2-singleValue"
                          >
                            Material Sample
                          </div>
                          <div
                            class=" css-1h01tm3-Input"
                            data-value=""
                          >
                            <input
                              aria-activedescendant=""
                              aria-autocomplete="list"
                              aria-expanded="false"
                              aria-haspopup="true"
                              autocapitalize="none"
                              autocomplete="off"
                              autocorrect="off"
                              class=""
                              disabled=""
                              id="react-select-2-input"
                              role="combobox"
                              spellcheck="false"
                              style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                              tabindex="0"
                              type="text"
                              value=""
                            />
                          </div>
                        </div>
                        <div
                          class=" css-1hb7zxy-IndicatorsContainer"
                        >
                          <span
                            class=" css-894a34-indicatorSeparator"
                          />
                          <div
                            aria-hidden="true"
                            class=" css-1xc3v61-indicatorContainer"
                          >
                            <svg
                              aria-hidden="true"
                              class="css-tj5bde-Svg"
                              focusable="false"
                              height="20"
                              viewBox="0 0 20 20"
                              width="20"
                            >
                              <path
                                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                              />
                            </svg>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </label>
              </div>
            </div>
          </div>
        </div>
        <h4
          class="mt-4"
        >
          Columns To Generate
        </h4>
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <strong>
              Add a new column:
            </strong>
            <div
              class="flex-fill mb-0"
            >
              <label
                class="insertField-field mb-2 w-100 mb-3"
              >
                <div
                  class="field-label label-col mb-2"
                />
                <div
                  class="field-col"
                  style="cursor: auto;"
                >
                  <div
                    class=""
                  >
                    <div
                      class=" css-b62m3t-container"
                    >
                      <span
                        class="css-1f43avz-a11yText-A11yText"
                        id="react-select-3-live-region"
                      />
                      <span
                        aria-atomic="false"
                        aria-live="polite"
                        aria-relevant="additions text"
                        class="css-1f43avz-a11yText-A11yText"
                        role="log"
                      />
                      <div
                        class="react-select__control css-13cymwt-control"
                      >
                        <div
                          class="react-select__value-container css-1fdsijx-ValueContainer"
                        >
                          <div
                            class="react-select__placeholder css-ef8t06-placeholder"
                            id="react-select-3-placeholder"
                          >
                            Select...
                          </div>
                          <div
                            class="react-select__input-container css-qbdosj-Input"
                            data-value=""
                          >
                            <input
                              aria-activedescendant=""
                              aria-autocomplete="list"
                              aria-describedby="react-select-3-placeholder"
                              aria-expanded="false"
                              aria-haspopup="true"
                              autocapitalize="none"
                              autocomplete="off"
                              autocorrect="off"
                              class="react-select__input"
                              id="react-select-3-input"
                              role="combobox"
                              spellcheck="false"
                              style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                              tabindex="0"
                              type="text"
                              value=""
                            />
                          </div>
                        </div>
                        <div
                          class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
                        >
                          <span
                            class="react-select__indicator-separator css-1u9des2-indicatorSeparator"
                          />
                          <div
                            aria-hidden="true"
                            class="react-select__indicator react-select__dropdown-indicator css-1xc3v61-indicatorContainer"
                          >
                            <svg
                              aria-hidden="true"
                              class="css-tj5bde-Svg"
                              focusable="false"
                              height="20"
                              viewBox="0 0 20 20"
                              width="20"
                            >
                              <path
                                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                              />
                            </svg>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </label>
            </div>
            <div
              class="d-grid"
            >
              <button
                class="btn btn-primary btn btn-primary"
                disabled=""
                type="button"
              >
                Add Column
              </button>
            </div>
          </div>
        </div>
        <h4
          class="mt-4"
        >
          Existing Templates
        </h4>
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <div>
              <div
                aria-label="Query Table"
                class="query-table-wrapper"
                role="search"
              >
                <div
                  class="d-flex align-items-end mb-1"
                >
                  <span>
                    Total matched records: 0
                  </span>
                  <div
                    class="ms-auto"
                  >
                    <div>
                      <div
                        class="d-flex gap-3"
                      />
                    </div>
                    <div
                      class="ms-auto"
                    >
                      <span
                        class="m-2"
                      >
                        <span>
                          <img
                            alt="To sort by multiple columns: Click one column header to sort by that column, then press shift while you click an additional header."
                            aria-describedby="queryTableMultiSortExplanation"
                            src="/static/images/iconInformation.gif"
                            tabindex="0"
                          />
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  class="ReactTable -striped -highlight"
                  data-testid="ReactTable"
                >
                  <div
                    class="table-responsive-wrapper"
                  >
                    <table
                      class="w-100"
                    >
                      <thead>
                        <tr>
                          <th
                            class="-cursor-pointer"
                            colspan="1"
                            style="width: auto;"
                          >
                            <div
                              class="-cursor-pointer select-none column-header"
                            >
                              <span
                                class="d-flex align-items-center justify-content-center"
                              >
                                <div
                                  class="originalFilename-field-header"
                                >
                                  Original Filename
                                </div>
                              </span>
                            </div>
                          </th>
                          <th
                            class="-cursor-pointer"
                            colspan="1"
                            style="width: auto;"
                          >
                            <div
                              class="-cursor-pointer select-none column-header"
                            >
                              <span
                                class="d-flex align-items-center justify-content-center"
                              >
                                <div
                                  class="createdBy-field-header"
                                >
                                  Created By
                                </div>
                              </span>
                            </div>
                          </th>
                          <th
                            class="-cursor-pointer"
                            colspan="1"
                            style="width: auto;"
                          >
                            <div
                              class="-cursor-pointer select-none column-header"
                            >
                              <span
                                class="d-flex align-items-center justify-content-center"
                              >
                                <div
                                  class="createdOn-field-header"
                                >
                                  Created On
                                </div>
                                <svg
                                  class="-sort-asc"
                                  fill="currentColor"
                                  height="1em"
                                  stroke="currentColor"
                                  stroke-width="0"
                                  viewBox="0 0 448 512"
                                  width="1em"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
                                  />
                                </svg>
                              </span>
                            </div>
                          </th>
                          <th
                            class=""
                            colspan="1"
                            style="width: auto;"
                          >
                            <div
                              class="column-header"
                            >
                              <span
                                class="d-flex align-items-center justify-content-center"
                              />
                            </div>
                          </th>
                        </tr>
                      </thead>
                      <tbody>
                        <tr>
                          <td
                            class="text-center"
                            colspan="4"
                          >
                            <div
                              class="spinner-border undefined"
                              role="status"
                            >
                              <span
                                class="visually-hidden mb-3"
                              >
                                Loading...
                              </span>
                            </div>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
                <div
                  class="mt-2"
                >
                  <span>
                    Total matched records: 0
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </main>
      <footer
        class="mt-3"
        id="wb-info"
        style="z-index: 0;"
      >
        <div
          class="brand"
        >
          <div
            class="container-fluid centered"
          >
            <div
              class="row "
            >
              <nav
                class="col-md-10 ftr-urlt-lnk py-3"
              >
                <ul>
                  <li>
                    <a
                      href="https://www.canada.ca/en/contact.html"
                    >
                      Contact information
                    </a>
                  </li>
                  <li>
                    <a
                      href="https://www.canada.ca/en/transparency/terms.html"
                    >
                      Terms and conditions
                    </a>
                  </li>
                  <li>
                    <a
                      href="https://www.canada.ca/en/transparency/privacy.html"
                    >
                      Privacy
                    </a>
                  </li>
                </ul>
                <div />
              </nav>
              <div
                class="col-6 col-md-3 col-lg-2 text-end pt-2"
              >
                <img
                  alt="Symbol of the Government of Canada"
                  class="mt-4"
                  src="https://www.canada.ca/etc/designs/canada/cdts/gcweb/v4_0_32/assets/wmms-blk.svg"
                />
              </div>
            </div>
          </div>
        </div>
      </footer>
    </form>
  </div>
</DocumentFragment>
`;
